ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,772 @@
+        }
+    }
+
+    /// `update` updates the current store to the next offset which is the previous offset plus `vec.actual_len()`.
+    /// It deallocates `vec`. `chunk_size` remains the same.
+    ///
+    /// ## Panics
+    /// Panics when `segment_size` is `None`
+    fn update(&mut self) {
+        // compute how many records have been added
+        // since `actual_length` is imprecise (i.e. rounded up) for segments smaller than 256,
+        // subtract `1` to make sure that offset is set sufficiently small
+        self.offset += ((self.actual_len() - 1) << 8) / self.segment_size.unwrap();
+        self.vec = Vec::<Option<UnverifiedValues>>::new();
+    }
+
+    /// `actual_len` computes the index of the last actual element plus `1`
+    fn actual_len(&self) -> usize {
+        let mut actual_length = self.vec.len();
+        while actual_length > 0 && self.vec[actual_length - 1].is_none() {
+            actual_length -= 1;
+        }
+        actual_length
+    }
+
+    /// `is_empty` returns `true` if at least one segment has been added
+    fn is_empty(&self) -> bool {
+        self.vec.is_empty()
+    }
+
+    fn initialize_segment_size(&mut self, segment_size: usize) {
+        self.segment_size = Some(segment_size);
+    }
+
+    /// `allocate_vec` allocates `vec`. The length is determined by the amount
+    /// of records, i.e. `chunk_size` and the size of a record, i.e. `segment_size`.
+    /// ## Panics
+    /// Panics when `segment_size` is `None`.
+    fn allocate_vec(&mut self) {
+        // add 255 to round up
+        self.vec = vec![None; (self.chunk_size * self.segment_size.unwrap() + 255) >> 8];
+    }
+
+    /// `reset_offset` reallocates vector such that a smaller `RecordId` than `offset` can be stored
+    /// `new_offset` will be the new `offset` and the store and store elements with `RecordId` `new_offset` and larger
+    /// when `new_offset` is larger that the current `offset`, the function does nothing
+    fn reset_offset(&mut self, new_offset: RecordId, segment_length: usize) {
+        if new_offset < self.offset {
+            let mut new_offset = usize::from(new_offset);
+            // if segment_length is less than 256, redefine new offset such that we don't need to rearrange the existing segments
+            if 256 % segment_length == 0 {
+                // distance between new_offset and old offset is rounded up to a multiple of 256
+                new_offset = ((usize::from(self.offset) - new_offset + 255) >> 8) << 8;
+            }
+            let extension_length = ((usize::from(self.offset) - new_offset) >> 8) * segment_length;
+            // use existing vec and add enough space in front
+            let vec_previous = &mut self.vec;
+            let mut vec: Vec<Option<UnverifiedValues>> =
+                Vec::with_capacity(vec_previous.len() + extension_length);
+            vec.resize(extension_length, None);
+            vec.append(vec_previous);
```

Comment:
I think you could instead do something like `self.vec.splice(0..0, repeat(None).take(extension_length))`.

(The current code is missing a `self.vec = vec` at the end -- also in `extend_vec`.)